src/scales/scale.time.js

```diff
@@ -257,38 +443,31 @@ module.exports = function(Chart) {
 			var ticksOpts = me.options.ticks;
 			var formats = timeOpts.displayFormats;
 			var capacity = me.getLabelCapacity(min);
-			var unit = timeOpts.unit || timeHelpers.determineUnit(timeOpts.minUnit, min, max, capacity);
-			var majorUnit = timeHelpers.determineMajorUnit(unit);
+			var unit = timeOpts.unit || determineUnit(timeOpts.minUnit, min, max, capacity);
+			var majorUnit = determineMajorUnit(unit);
+			var timestamps = [];
 			var ticks = [];
 			var i, ilen, timestamp, stepSize;
 
-			if (ticksOpts.source === 'labels') {
-				for (i = 0, ilen = model.labels.length; i < ilen; ++i) {
-					timestamp = model.labels[i];
-					if (timestamp >= min && timestamp <= max) {
-						ticks.push(timestamp);
-					}
-				}
-			} else {
+			if (ticksOpts.source === 'auto') {
 				stepSize = helpers.valueOrDefault(timeOpts.stepSize, timeOpts.unitStepSize)
-					|| timeHelpers.determineStepSize(min, max, unit, capacity);
-
-				ticks = timeHelpers.generateTicks({
-					maxTicks: capacity,
-					min: parse(timeOpts.min, me),
-					max: parse(timeOpts.max, me),
-					stepSize: stepSize,
-					majorUnit: majorUnit,
-					unit: unit,
-					timeOpts: timeOpts
-				}, {
-					min: min,
-					max: max
-				});
-
-				// Recompute min/max, the ticks generation might have changed them (BUG?)
-				min = ticks.length ? ticks[0] : min;
-				max = ticks.length ? ticks[ticks.length - 1] : max;
+					|| determineStepSize(min, max, unit, capacity);
+
+				timestamps = generate(min, max, unit, majorUnit, stepSize, timeOpts);
+
+				// Expand min/max to the generated ticks
+				min = helpers.isNullOrUndef(timeOpts.min) && timestamps.length ? timestamps[0] : min;
+				max = helpers.isNullOrUndef(timeOpts.max) && timestamps.length ? timestamps[timestamps.length - 1] : max;
+			} else {
+				timestamps = model.labels;
+			}
+
+			// Remove ticks outside the min/max range
```

Comment:
I think it'd be slightly cleaner if this were able to be moved inside `generate`. That would have a few benefits like being able to revert the test change so that the test is more meaningful and not needing the comment above `generate` explaining that it might not do exactly what might be expected